File: data_structures/cpp/queues.cpp
#include <iostream>
#include <vector>

template <typename T>
class CustomQueue {
private:
    std::vector<T> elements;

public:
    // MARK: - Queue Operations

    /// Enqueue an element to the back of the queue.
    void enqueue(T element) {
        elements.push_back(element);
    }

    /// Dequeue an element from the front of the queue.
    T dequeue() {
        if (elements.empty()) {
            throw std::out_of_range("Queue is empty.");
        }
        T frontElement = elements.front();
        elements.erase(elements.begin());
        return frontElement;
    }

    /// Peek at the front element in the queue without dequeuing.
    T peek() const {
        if (elements.empty()) {
            throw std::out_of_range("Queue is empty.");
        }
        return elements.front();
    }

    /// Check if the queue is empty.
    bool isEmpty() const {
        return elements.empty();
    }

    /// Get the number of elements in the queue.
    int count() const {
        return elements.size();
    }

    // MARK: - Queue Operations with Array

    /// Initialize the queue with an array of elements.
    void initializeWithArray(std::vector<T> array) {
        elements = array;
    }

    /// Enqueue an array of elements to the back of the queue.
    void enqueueArray(const std::vector<T>& array) {
        elements.insert(elements.end(), array.begin(), array.end());
    }

    /// Dequeue an array of elements from the front of the queue.
    std::vector<T> dequeueArray(int count) {
        if (count < 0 || count > elements.size()) {
            throw std::out_of_range("Invalid count for dequeueArray.");
        }

        std::vector<T> dequeuedElements;
        for (int i = 0; i < count; ++i) {
            dequeuedElements.push_back(elements.front());
            elements.erase(elements.begin());
        }
        return dequeuedElements;
    }

    // MARK: - Clear and Reset

    /// Remove all elements from the queue.
    void clear() {
        elements.clear();
    }

    // MARK: - Checking for Element

    /// Check if the queue contains a specific element.
    bool contains(T element) const {
        return std::find(elements.begin(), elements.end(), element) != elements.end();
    }

    // MARK: - Filtering

    /// Filter the queue using a given predicate.
    std::vector<T> filter(bool (*predicate)(T)) const {
        std::vector<T> filtered;
        for (const T& element : elements) {
            if (predicate(element)) {
                filtered.push_back(element);
            }
        }
        return filtered;
    }

    // MARK: - Conversion to Vector

    /// Convert the queue to a vector.
    std::vector<T> toVector() const {
        return elements;
    }

    // MARK: - Map

    /// Transform each element in the queue using a provided transform function.
    template <typename U>
    std::vector<U> map(U (*transform)(T)) const {
        std::vector<U> mapped;
        for (const T& element : elements) {
            mapped.push_back(transform(element));
        }
        return mapped;
    }

    // MARK: - Reduce

    /// Combine all elements in the queue using a reducer function.
    template <typename Result>
    Result reduce(Result initialResult, Result (*reducer)(Result, T)) const {
        Result result = initialResult;
        for (const T& element : elements) {
            result = reducer(result, element);
        }
        return result;
    }

    // MARK: - Concatenate

    /// Concatenate another queue to this queue.
    void concatenate(const CustomQueue<T>& otherQueue) {
        elements.insert(elements.end(), otherQueue.elements.begin(), otherQueue.elements.end());
    }

    // MARK: - Subscript

    /// Access the element at a specific index in the queue.
    T operator[](int index) const {
        if (index >= 0 && index < elements.size()) {
            return elements[index];
        }
        throw std::out_of_range("Index out of range.");
    }

    // ... Add more queue operations as needed ...
};

int main() {
    CustomQueue<int> queue;
    queue.enqueue(1);
    queue.enqueue(2);
    queue.enqueue(3);

    std::cout << "Queue:" << std::endl;
    while (!queue.isEmpty()) {
        std::cout << queue.dequeue() << std::endl;
    }

    return 0;
}

File: data_structures/rust/queues.rs
pub struct CustomQueue<T> {
    elements: Vec<T>,
}

impl<T> CustomQueue<T> {
    // MARK: - Queue Operations

    /// Enqueue an element to the back of the queue.
    pub fn enqueue(&mut self, element: T) {
        self.elements.push(element);
    }

    /// Dequeue an element from the front of the queue.
    pub fn dequeue(&mut self) -> Option<T> {
        self.elements.pop()
    }

    /// Peek at the front element in the queue without dequeuing.
    pub fn peek(&self) -> Option<&T> {
        self.elements.first()
    }

    /// Check if the queue is empty.
    pub fn is_empty(&self) -> bool {
        self.elements.is_empty()
    }

    /// Get the number of elements in the queue.
    pub fn count(&self) -> usize {
        self.elements.len()
    }

    // MARK: - Queue Operations with Vector

    /// Initialize the queue with a vector of elements.
    pub fn initialize_with_vector(&mut self, vec: Vec<T>) {
        self.elements = vec;
    }

    /// Enqueue a vector of elements to the back of the queue.
    pub fn enqueue_vector(&mut self, vec: Vec<T>) {
        self.elements.extend(vec);
    }

    /// Dequeue a specified number of elements from the front of the queue.
    pub fn dequeue_vector(&mut self, count: usize) -> Vec<T> {
        if count == 0 {
            return Vec::new();
        }
        let dequeued_elements: Vec<T> = self.elements.drain(0..count).collect();
        dequeued_elements
    }

    // MARK: - Clear and Reset

    /// Remove all elements from the queue.
    pub fn clear(&mut self) {
        self.elements.clear();
    }

    // MARK: - Checking for Element

    /// Check if the queue contains a specific element.
    pub fn contains(&self, element: &T) -> bool
    where
        T: PartialEq,
    {
        self.elements.contains(element)
    }

    // MARK: - Filtering

    /// Filter the queue using a given predicate.
    pub fn filter<F>(&self, predicate: F) -> Vec<T>
    where
        F: Fn(&T) -> bool,
    {
        self.elements.iter().filter(|&x| predicate(x)).cloned().collect()
    }

    // MARK: - Conversion to Vector

    /// Convert the queue to a vector.
    pub fn to_vector(&self) -> Vec<T> {
        self.elements.clone()
    }

    // MARK: - Map

    /// Transform each element in the queue using a provided transform function.
    pub fn map<U, F>(&self, transform: F) -> Vec<U>
    where
        F: Fn(&T) -> U,
    {
        self.elements.iter().map(transform).collect()
    }

    // MARK: - Reduce

    /// Combine all elements in the queue using a reducer function.
    pub fn reduce<Result, F>(&self, initial_result: Result, reducer: F) -> Result
    where
        F: Fn(Result, &T) -> Result,
    {
        self.elements.iter().fold(initial_result, |acc, x| reducer(acc, x))
    }

    // MARK: - Concatenate

    /// Concatenate another queue to this queue.
    pub fn concatenate(&mut self, other_queue: CustomQueue<T>) {
        self.elements.extend(other_queue.elements);
    }

    // MARK: - Subscript

    /// Access the element at a specific index in the queue.
    pub fn get(&self, index: usize) -> Option<&T> {
        self.elements.get(index)
    }

    // ... Add more queue operations as needed ...
}

fn main() {
    let mut queue = CustomQueue {
        elements: Vec::new(),
    };
    queue.enqueue(1);
    queue.enqueue(2);
    queue.enqueue(3);

    println!("Queue:");
    while !queue.is_empty() {
        if let Some(element) = queue.dequeue() {
            println!("{}", element);
        }
    }
}
